lib/main.dart

```diff
@@ -105,3 +113,58 @@ void main() async {
     }
   });
 }
+
+Future<void> _onBackgroundMessage(RemoteMessage message) async {
+  print("Handling a background message: ${message.messageId}\nMessage data: ${message.data}");
+  await initializeBreezServices();
+  await Workmanager().initialize(callbackDispatcher, isInDebugMode: true);
+  var data = message.data["data"] ?? message.data;
+  if (data != null) {
+    // This is required to retrieve payment_hash from notification payload
+    // TODO: Payload's format should be changed on server-side so this information is accessible from RemoteMessage without deserialization
+    if (data is String) data = json.decode(data);
+  }
+
+  switch (message.data["notification_type"]) {
+    case "payment_received":
+      await Workmanager().registerOneOffTask(
+        data["payment_hash"],
+        message.data["notification_type"], // Ignored on iOS
+        constraints: Constraints(networkType: NetworkType.connected),
+        inputData: message.data, // We need to parse taskName from inputData as taskName is ignored on iOS
+      );
+      break;
+  }
+
+  return Future<void>.value();
+}
+
+@pragma('vm:entry-point') // Mandatory if the App is obfuscated or using Flutter 3.1+
+void callbackDispatcher() {
+  const timeoutDuration = Duration(seconds: 60);
+
+  Workmanager().executeTask((task, inputData) async {
+    final taskCompleter = Completer<bool>();
+
+    if (inputData != null) {
+      var data = inputData["data"] ?? inputData;
+      if (data != null) {
+        if (data is String) data = json.decode(data);
+      }
+
+      switch (inputData["notification_type"]) {
+        case "payment_received":
+          pollForReceivedPayment(data, taskCompleter);
+          break;
+      }
+    }
+
+    return taskCompleter.future.timeout(
```

Comment:
We should also exit early if the payment is received.

Reply:
We do complete the taskCompleter when there's a match.